src/Bicep.Core/TypeSystem/DeployTimeConstantVisitor.cs

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+using Bicep.Core.Diagnostics;
+using Bicep.Core.SemanticModel;
+using Bicep.Core.Syntax;
+
+namespace Bicep.Core.TypeSystem
+{
+    /// <summary>
+    /// Visitor used to collect errors caused by property assignments to run-time variables,
+    /// currently, we only run this on the "name" property of resources and modules
+    /// </summary>
+    public sealed class DeployTimeConstantVisitor : SyntaxVisitor
+    {
+
+        private readonly SemanticModel.SemanticModel model;
+        private readonly IDiagnosticWriter diagnosticWriter;
+
+        // used for logging the property when we detect a run time property
+        private string currentSymbol;
+        // whether a runtimeValue is allowed for the currentSymbol
+        private bool runtimeValueAllowed;
+
+        private DeployTimeConstantVisitor(SemanticModel.SemanticModel model, IDiagnosticWriter diagnosticWriter)
+        {
+            this.model = model;
+            this.diagnosticWriter = diagnosticWriter;
+            this.runtimeValueAllowed = true;
+            this.currentSymbol = "";
+        }
+
+        // entry point for this visitor
+        public static void ValidateDeployTimeConstants(SemanticModel.SemanticModel model, IDiagnosticWriter diagnosticWriter)
+        {
+            new DeployTimeConstantVisitor(model, diagnosticWriter).Visit(model.Root.Syntax);
+        }
+
+        public override void VisitResourceDeclarationSyntax(ResourceDeclarationSyntax syntax)
+        {
+            this.runtimeValueAllowed = false;
+            base.VisitResourceDeclarationSyntax(syntax);
+            this.runtimeValueAllowed = true;
+        }
+
+        public override void VisitModuleDeclarationSyntax(ModuleDeclarationSyntax syntax)
+        {
+            this.runtimeValueAllowed = false;
+            base.VisitModuleDeclarationSyntax(syntax);
+            this.runtimeValueAllowed = true;
+        }
+
+        public override void VisitObjectPropertySyntax(ObjectPropertySyntax syntax)
```

Comment:
Should we ensure the name identifier is only at the top level? It is possible that a resource contains a `name` property in tags.

Reply:
I never explicitly thought about this but the current approach can handle this. In VisitObjectPropertySyntax we set the reset `runtimeValueAllowed` property according to name. So the name `tags` would set the `runtimeValueAllowed` in its children to `true`. Therefore, it doesn't matter if we have a tag named "name"

Reply:
When visiting the `name` tag, will it set `runtimeValueAllowed` back to `false` again?

Reply:
Nvm it's guarded by `!this.runtimeValueAllowed`.